diff --git a/build_settings.cc b/build_settings.cc
index 4cf2b86..d349c06 100644
--- a/build_settings.cc
+++ b/build_settings.cc
@@ -7,6 +7,7 @@
 #include <utility>
 
 #include "base/files/file_util.h"
+#include "base/path_service.h"
 #include "tools/gn/filesystem_utils.h"
 
 BuildSettings::BuildSettings() {
@@ -18,6 +19,8 @@ BuildSettings::BuildSettings(const BuildSettings& other)
       secondary_source_path_(other.secondary_source_path_),
       python_path_(other.python_path_),
       build_config_file_(other.build_config_file_),
+      build_config_root_path_(other.build_config_root_path_),
+      build_config_root_path_utf8_(other.build_config_root_path_utf8_),
       arg_file_template_path_(other.arg_file_template_path_),
       build_dir_(other.build_dir_),
       build_args_(other.build_args_) {}
@@ -35,6 +38,20 @@ void BuildSettings::SetRootPath(const base::FilePath& r) {
   root_path_utf8_ = FilePathToUTF8(root_path_);
 }
 
+void BuildSettings::SetBuildConfigRootPath(const base::FilePath& r) {
+  DCHECK(r.value()[r.value().size() - 1] != base::FilePath::kSeparators[0]);
+  if (r == base::FilePath("GN_INSTALL_PATH"))
+  {
+    PathService::Get(base::DIR_EXE, &build_config_root_path_);
+    build_config_root_path_utf8_ = FilePathToUTF8(build_config_root_path_);
+  }
+  else
+  {
+    build_config_root_path_ = r.NormalizePathSeparatorsTo('/');
+    build_config_root_path_utf8_ = FilePathToUTF8(build_config_root_path_);
+  }
+}
+
 void BuildSettings::SetSecondarySourcePath(const SourceDir& d) {
   secondary_source_path_ = GetFullPath(d).NormalizePathSeparatorsTo('/');
 }
@@ -61,6 +78,16 @@ base::FilePath BuildSettings::GetFullPathSecondary(
   return dir.Resolve(secondary_source_path_).NormalizePathSeparatorsTo('/');
 }
 
+base::FilePath BuildSettings::GetFullPathBuildConfig(const SourceFile& file) const
+{
+  return file.Resolve(build_config_root_path_).NormalizePathSeparatorsTo('/');
+}
+
+base::FilePath BuildSettings::GetFullPathBuildConfig(const SourceDir& dir) const
+{
+  return dir.Resolve(build_config_root_path_).NormalizePathSeparatorsTo('/');
+}
+
 void BuildSettings::ItemDefined(std::unique_ptr<Item> item) const {
   DCHECK(item);
   if (!item_defined_callback_.is_null())
diff --git a/build_settings.h b/build_settings.h
index 670f996..6bf0d44 100644
--- a/build_settings.h
+++ b/build_settings.h
@@ -57,6 +57,10 @@ class BuildSettings {
   const SourceFile& build_config_file() const { return build_config_file_; }
   void set_build_config_file(const SourceFile& f) { build_config_file_ = f; }
 
+  const base::FilePath& build_config_root_path() const { return build_config_root_path_; }
+  const std::string& build_config_root_path_utf8() const { return build_config_root_path_utf8_; }
+  void SetBuildConfigRootPath(const base::FilePath& r);
+
   // Path to a file containing the default text to use when running `gn args`.
   const SourceFile& arg_file_template_path() const {
     return arg_file_template_path_;
@@ -86,6 +90,9 @@ class BuildSettings {
   base::FilePath GetFullPathSecondary(const SourceFile& file) const;
   base::FilePath GetFullPathSecondary(const SourceDir& dir) const;
 
+  base::FilePath GetFullPathBuildConfig(const SourceFile& file) const;
+  base::FilePath GetFullPathBuildConfig(const SourceDir& dir) const;
+
   // Called when an item is defined from a background thread.
   void ItemDefined(std::unique_ptr<Item> item) const;
   void set_item_defined_callback(ItemDefinedCallback cb) {
@@ -116,6 +123,8 @@ class BuildSettings {
   base::FilePath python_path_;
 
   SourceFile build_config_file_;
+  base::FilePath build_config_root_path_;
+  std::string build_config_root_path_utf8_;
   SourceFile arg_file_template_path_;
   SourceDir build_dir_;
   Args build_args_;
diff --git a/config.cc b/config.cc
index fcc5fa6..119b1c8 100644
--- a/config.cc
+++ b/config.cc
@@ -26,6 +26,16 @@ bool Config::OnResolved(Err* err) {
   DCHECK(!resolved_);
   resolved_ = true;
 
+  if (!settings()->build_settings()->build_config_root_path().empty() &&
+       label().name() == "default_include_dirs") {
+     std::string filepath =
+         settings()->build_settings()->build_config_root_path_utf8();
+ #if defined(OS_WIN)
+     filepath = "/" + filepath;
+ #endif
+     own_values_.include_dirs().push_back(SourceDir(filepath));
+  }
+
   if (!configs_.empty()) {
     // Subconfigs, flatten.
     //
diff --git a/function_exec_script.cc b/function_exec_script.cc
index 8c4909c..34282cd 100644
--- a/function_exec_script.cc
+++ b/function_exec_script.cc
@@ -139,6 +139,10 @@ Value RunExecScript(Scope* scope,
   if (err->has_error())
     return Value();
   base::FilePath script_path = build_settings->GetFullPath(script_source);
+  if (!base::PathExists(script_path))
+  {
+    script_path = build_settings->GetFullPathBuildConfig(script_source);
+  }
   if (!build_settings->secondary_source_path().empty() &&
       !base::PathExists(script_path)) {
     // Fall back to secondary source root when the file doesn't exist.
diff --git a/function_rebase_path.cc b/function_rebase_path.cc
index 7626551..7aaaf22 100644
--- a/function_rebase_path.cc
+++ b/function_rebase_path.cc
@@ -4,6 +4,8 @@
 
 #include <stddef.h>
 
+#include "base/strings/string_util.h"
+#include "base/files/file_util.h"
 #include "tools/gn/build_settings.h"
 #include "tools/gn/filesystem_utils.h"
 #include "tools/gn/functions.h"
@@ -69,16 +71,28 @@ Value ConvertOnePath(const Scope* scope,
   bool looks_like_dir = ValueLooksLikeDir(string_value);
 
   // System-absolute output special case.
-  if (convert_to_system_absolute) {
+  if (convert_to_system_absolute || !scope->settings()->build_settings()->build_config_root_path().empty()) {
     base::FilePath system_path;
     if (looks_like_dir) {
       system_path = scope->settings()->build_settings()->GetFullPath(
+        from_dir.ResolveRelativeDir(value, err,
+          scope->settings()->build_settings()->root_path_utf8()));
+      if (!scope->settings()->build_settings()->build_config_root_path().empty() &&
+          !base::PathExists(system_path)) {
+        system_path = scope->settings()->build_settings()->GetFullPathBuildConfig(
           from_dir.ResolveRelativeDir(value, err,
-              scope->settings()->build_settings()->root_path_utf8()));
+            scope->settings()->build_settings()->build_config_root_path_utf8()));
+      }
     } else {
       system_path = scope->settings()->build_settings()->GetFullPath(
           from_dir.ResolveRelativeFile(value, err,
               scope->settings()->build_settings()->root_path_utf8()));
+      if (!scope->settings()->build_settings()->build_config_root_path().empty() &&
+          !base::PathExists(system_path)) {
+        system_path = scope->settings()->build_settings()->GetFullPathBuildConfig(
+          from_dir.ResolveRelativeFile(value, err,
+            scope->settings()->build_settings()->build_config_root_path_utf8()));
+      }
     }
     if (err->has_error())
       return Value();
diff --git a/input_file_manager.cc b/input_file_manager.cc
index 9fd3def..3b6f622 100644
--- a/input_file_manager.cc
+++ b/input_file_manager.cc
@@ -42,19 +42,40 @@ bool DoLoadFile(const LocationRange& origin,
   base::FilePath primary_path = build_settings->GetFullPath(name);
   ScopedTrace load_trace(TraceItem::TRACE_FILE_LOAD, name.value());
   if (!file->Load(primary_path)) {
+    bool found_in_path = false;
     if (!build_settings->secondary_source_path().empty()) {
       // Fall back to secondary source tree.
       base::FilePath secondary_path =
-          build_settings->GetFullPathSecondary(name);
-      if (!file->Load(secondary_path)) {
+        build_settings->GetFullPathSecondary(name);
+      if (!file->Load(secondary_path) && !build_settings->build_config_root_path().empty()) {
         *err = Err(origin, "Can't load input file.",
                    "Unable to load:\n  " +
-                   FilePathToUTF8(primary_path) + "\n"
-                   "I also checked in the secondary tree for:\n  " +
-                   FilePathToUTF8(secondary_path));
+                    FilePathToUTF8(primary_path) + "\n"
+                    "I also checked in the secondary tree for:\n  " +
+                    FilePathToUTF8(secondary_path));
         return false;
       }
-    } else {
+
+      found_in_path = true;
+    }
+
+    // Fall back to build config root
+    if (!found_in_path && !build_settings->build_config_root_path().empty()) {
+      if (!file->Load(build_settings->GetFullPathBuildConfig(name)))
+      {
+        *err = Err(origin, "Can't load input file.",
+                   "Unable to load:\n  " +
+                    FilePathToUTF8(primary_path) + "\n"
+                    "I also checked in the build config tree for:\n  " +
+                    FilePathToUTF8(build_settings->build_config_root_path()));
+        return false;
+      }
+
+      found_in_path = true;  // build_config_root_path
+    }
+
+    if (!found_in_path)
+    {
       *err = Err(origin,
                  "Unable to load \"" + FilePathToUTF8(primary_path) + "\".");
       return false;
diff --git a/setup.cc b/setup.cc
index 7fc4a2e..a2b932a 100644
--- a/setup.cc
+++ b/setup.cc
@@ -746,7 +746,19 @@ bool Setup::FillOtherConfig(const base::CommandLine& cmdline) {
     return false;
   }
   build_settings_.set_build_config_file(
-      SourceFile(build_config_value->string_value()));
+    SourceFile(build_config_value->string_value()));
+
+  const Value* build_config_root_value =
+  dotfile_scope_.GetValue("buildconfig_root", true);
+  if (build_config_root_value) {
+    if (!build_config_root_value->VerifyTypeIs(Value::STRING, &err)) {
+      err.PrintToStdout();
+      return false;
+    }
+
+    build_settings_.SetBuildConfigRootPath(
+      UTF8ToFilePath(build_config_root_value->string_value()));
+  }
 
   // Targets to check.
   const Value* check_targets_value =
diff --git a/value_extractors.cc b/value_extractors.cc
index ff009ce..4b56fbb 100644
--- a/value_extractors.cc
+++ b/value_extractors.cc
@@ -6,8 +6,11 @@
 
 #include <stddef.h>
 
+#include "base/strings/string_util.h"
+#include "base/files/file_util.h"
 #include "tools/gn/build_settings.h"
 #include "tools/gn/err.h"
+#include "tools/gn/filesystem_utils.h"
 #include "tools/gn/label.h"
 #include "tools/gn/source_dir.h"
 #include "tools/gn/source_file.h"
@@ -67,8 +70,18 @@ struct RelativeFileConverter {
         current_dir(current_dir_in) {
   }
   bool operator()(const Value& v, SourceFile* out, Err* err) const {
-    *out = current_dir.ResolveRelativeFile(v, err,
-                                           build_settings->root_path_utf8());
+    *out = current_dir.ResolveRelativeFile(
+        v, err, build_settings->root_path_utf8());
+    if (!build_settings->build_config_root_path().empty()) {
+       base::FilePath rf_path = build_settings->GetFullPath(*out);
+       if (!base::PathExists(rf_path)) {
+         std::string p = FilePathToUTF8(build_settings->GetFullPathBuildConfig(*out));
+ #if defined(OS_WIN)
+         p = "/" + p;
+ #endif
+         *out = SourceFile(p);
+       }
+    }
     return !err->has_error();
   }
   const BuildSettings* build_settings;
